cpp/src/arrow/compute/exec/spilling_util.cc

```diff
@@ -0,0 +1,483 @@
+    array_info.num_children = array->child_data.size();
+    for(size_t i = 0; i < array->buffers.size(); i++)
+    {
+        if(array->buffers[i])
+        {
+            array_info.sizes[i] = array->buffers[i]->size();
+            total_size += array_info.sizes[i];
+            uintptr_t addr = array->buffers[i]->address();
+            if((addr % SpillFile::kAlignment) != 0)
+                return Status::Invalid("Buffer not aligned to 512 bytes!");
+            array_info.bufs[i] = std::move(array->buffers[i]);
+#ifndef _WIN32
+            struct iovec io;
+            io.iov_base = static_cast<void *>(array_info.bufs[i]->mutable_data());
+            io.iov_len = static_cast<size_t>(array_info.sizes[i]);
+            batch_info.ios.push_back(io);
+#endif
+        }
+    }
+
+    array_info.dictionary = std::move(array->dictionary);
+    for(std::shared_ptr<ArrayData> &child : array->child_data)
+        RETURN_NOT_OK(CollectArrayInfo(batch_info, total_size, child.get()));
+
+    // Cleanup the ArrayData
+    array->type.reset();
+    array->length = 0;
+    return Status::OK();
+}
+
+static Status AllocateBuffersForBatch(SpillFile::BatchInfo &batch_info, MemoryPool *pool)
+{
+#ifndef _WIN32
+    size_t iiovec = 0;
+#endif
+    for(ArrayInfo &arr : batch_info.arrays)
+    {
+        for(size_t ibuf = 0; ibuf < 3; ibuf++)
+        {
+            if(arr.sizes[ibuf])
+            {
+                ARROW_ASSIGN_OR_RAISE(
+                    arr.bufs[ibuf],
+                    AllocateBuffer(
+                        arr.sizes[ibuf],
+                        SpillFile::kAlignment, pool));
+#ifndef _WIN32
+                batch_info.ios[iiovec].iov_base = static_cast<void *>(arr.bufs[ibuf]->mutable_data());
+                batch_info.ios[iiovec].iov_len = static_cast<size_t>(arr.sizes[ibuf]);
+                iiovec++;
+#endif
+            }
+        }
+    }
+    return Status::OK();
+}
+
+Status SpillFile::SpillBatch(QueryContext *ctx, ExecBatch batch)
+{
+    if(handle_ == kInvalidHandle)
+    {
+        ARROW_ASSIGN_OR_RAISE(handle_, OpenTemporaryFile());
+    }
+    int64_t total_size = 0;
+    batches_.emplace_back(new BatchInfo);
+    BatchInfo *info = batches_.back();
+    for(int i = 0; i < batch.num_values(); i++)
+    {
+        if (batch[i].is_scalar())
+            return Status::Invalid("Cannot spill a Scalar");
+        RETURN_NOT_OK(CollectArrayInfo(*info, total_size, batch[i].mutable_array()));
+    }
+    info->start = size_;
+    size_ += total_size;
+
+    FileHandle handle = handle_;
+    RETURN_NOT_OK(ctx->ScheduleIOTask(
+                      [this, handle, info, ctx, total_size]()
+                      {
+                          auto mark = ctx->ReportTempFileIO(total_size);
+                          RETURN_NOT_OK(WriteBatch_PlatformSpecific(handle, *info));
+                          if(++batches_written_ == batches_.size() && read_requested_.load())
```

Comment:
It is not clear why this expression is correct. For instance it requires that batches_.size() cannot change concurrently. Similarly, the way read_requested_ is changed may also matter. Can you add a comment justifying correctness?